dbms/src/Server/StorageConfigParser.cpp

```diff
@@ -444,9 +446,135 @@ std::tuple<size_t, TiFlashStorageConfig> TiFlashStorageConfig::parseSettings(
         storage_config.remote_cache_config.parse(config.getString("storage.remote.cache"), log);
     }
 
+    if (config.has("storage.tmp"))
+    {
+        // Need to make sure storage.main/latest is parsed before storage.tmp.
+        storage_config.parseTmpConfig(config.getString("storage"));
+    }
+    else if (config.has("tmp_path"))
+    {
+        storage_config.tmp_path = getNormalizedPath(config.getString("tmp_path"));
+        // If storage.tmp doesn't exist, tmp_capacity will be zero.
+        storage_config.tmp_capacity = 0;
+    }
+    else
+    {
+        storage_config.tmp_path = getNormalizedPath(storage_config.latest_data_paths[0] + "tmp/");
+        storage_config.tmp_capacity = 0;
+    }
+    LOG_INFO(
+        log,
+        "storage.tmp config parse done: tmp_path: {}, tmp_capacity: {}",
+        storage_config.tmp_path,
+        storage_config.tmp_capacity);
+
     return std::make_tuple(global_capacity_quota, storage_config);
 }
 
+void TiFlashStorageConfig::parseTmpConfig(const String & content)
+{
+    std::istringstream ss(content);
+    cpptoml::parser p(ss);
+    auto table = p.parse();
+
+    const auto tmp_path_opt = table->get_qualified_as<String>("tmp.dir");
+    if (!tmp_path_opt || tmp_path_opt->empty())
+        tmp_path = latest_data_paths[0] + "tmp/";
+    else
+        tmp_path = *tmp_path_opt;
+    tmp_path = getNormalizedPath(tmp_path);
+
+    tmp_capacity = 0;
+    const auto tmp_capacity_opt = table->get_qualified_as<UInt64>("tmp.capacity");
+    if (tmp_capacity_opt)
+        tmp_capacity = *tmp_capacity_opt;
+}
+
+// Seperate this function from TiFlashStorageConfig::parseTmpConfig() because need to create tmp_path first.
+void TiFlashStorageConfig::checkTmpCapacity(UInt64 global_capacity_quota, const LoggerPtr & log) const
+{
+    // global_capacity_quota and storage.main/latest.capacity cannot take effects at the same time.
+    RUNTIME_CHECK(!(!main_capacity_quota.empty() && global_capacity_quota > 0));
+    RUNTIME_CHECK_MSG(
+        main_capacity_quota.size() == latest_capacity_quota.size(),
+        "main_capacity_quota.size: {}, latest_capacity_quota.size: {}",
+        main_capacity_quota.size(),
+        latest_capacity_quota.size());
```

Comment:
why `main_capacity_quota.size() == latest_capacity_quota.size()`?
The user could have 1 SSD for setting `storage.latest.{dir, capacity}` and multiple SATA SSD for setting `storage.main.{dir, capacity}`

Reply:
yes. you are right.